error-constructors.d.ts

```diff
@@ -1,8 +1,30 @@
-type BaseErrorConstructor = new (message?: string, ...arguments_: unknown[]) => Error;
+export type ErrorFactory<T extends Error> = () => T;
 
 /**
 Let `serialize-error` know about your custom error constructors so that when `{name: 'MyCustomError', message: 'It broke'}` is found, it uses the right error constructor. If "MyCustomError" isn't found in the global list of known constructors, it defaults to the base `Error` error constructor.
 
-Warning: The constructor must work without any arguments or this function will throw.
+@param constructor - The error constructor to add.
+@param factory - Optional factory function to create instances. Required if the constructor doesn't work without arguments.
+
+@example
+```
+// Simple constructor that works without arguments
+addKnownErrorConstructor(MyError);
+
+// Constructor that requires arguments with factory function
+class CustomError extends Error {
+	constructor(message, options = {}) {
+		super(message);
+		this.name = 'CustomError';
+		this.code = options.code ?? 'UNKNOWN';
+	}
+}
+
+// Use a factory function to provide default arguments
+addKnownErrorConstructor(CustomError, () => new CustomError('', {code: 'ERR_UNICORN'}));
+```
 */
-export function addKnownErrorConstructor(constructor: BaseErrorConstructor): void;
+export function addKnownErrorConstructor<T extends Error>(
+	constructor: new (...arguments_: any[]) => T,
+	factory?: ErrorFactory<T>
```

Comment:
Would be great to make this type tighter so you're forced to provide a factory when the first argument is not a string, for example. If it is, it can remain optional